PSF Fitting accounts for corners

amn3142 opened the pull request:
Lots of extra comments to try to help explain my thought process...
Test function right now only checks whether uniform kernels are returned as a final uniform kernel for a symmetry of 6.

sibirrer commented on lenstronomy/Workflow/psf_fitting.py:
```diff
@@ -129,6 +130,15 @@ def update_psf(self, kwargs_psf, kwargs_params, stacking_method='median', psf_sy
         :param new_procedure: boolean, uses post lenstronomy 1.9.2 procedure which is more optimal for super-sampled
          PSF's
+
+         :boolean corner_symmetry: if the imposed symmetry is an odd number, the edges of the reconstructed PSF in its default form will be
```
follow the same documentation style with :param corner_symmetry:

sibirrer replied:
:param corner_symmetry:

sibirrer commented on lenstronomy/Workflow/psf_fitting.py:
```diff
@@ -179,7 +190,8 @@ def update_psf(self, kwargs_psf, kwargs_params, stacking_method='median', psf_sy
         else:
-            kernel_old_high_res = psf_class.kernel_point_source_supersampled(supersampling_factor=point_source_supersampling_factor)
+            kernel_old_high_res = psf_class.kernel_point_source_supersampled(
+                supersampling_factor=point_source_supersampling_factor)
```
this is weird formatting. You might not need a line break here

sibirrer commented on lenstronomy/Workflow/psf_fitting.py:
```diff
@@ -409,36 +426,92 @@ def combine_psf(kernel_list_new, kernel_old, factor=1., stacking_option='median'
+        :return: updated PSF estimate
+        """
+        ##keep_corners is a boolean which tracks whether to calc PSF separately in the corners for odd symmetry rotations.
+        keep_corners = type(corner_symmetry) == int
```
I don't understand this line. Isn't corner_symmetry supposed to be a boolean already?

amn3142 replied:
typo... corner_symmetry should be an int or None, fixed.

sibirrer reviewed (changes requested):
This looks good to me. I have some minor documentation format requests but otherwise ready to go. Thank you very much!

sibirrer commented:
> Hi @sibirrer I realized it didn't make sense to calculate the corners mask potentially thousands of times, so I've moved the corner completeness calculation to before the iterative part of the psf fitting process. This changed some of the structure significantly, including adding a mask calculation function.
> 
> It all seems to work, however, I didn't think of a good way to actually use the test functions to check that the correct information is getting passed through update_iterative (other than on my own with print statements). I wanted to make sure this updated structure would be ok for you before I go through the other comments you had.

That sounds good @amn3142 ! For the test, you can simply check the individual function, no need for 'system-integrated' testing with dedicated outputs. Thanks a lot!

amn3142 commented on lenstronomy/Workflow/psf_fitting.py:
```diff
@@ -399,7 +414,9 @@ def cutout_psf_single(x, y, image, mask, kernel_size, kernel_init):
-    def combine_psf(kernel_list_new, kernel_old, factor=1., stacking_option='median', symmetry=1):
+    ####Correct this based on Maverick Oh's note about lack of flux conservation.
+    def combine_psf(kernel_list_new, kernel_old, factor=1., stacking_option='median', symmetry=1, corner_symmetry=None):
+        ##TODO: Account for image edges/masked pixels.
```
This one is. Right now, we don't do any weighting based on missing information in the data (e.g. image edge or bad pixels).

amn3142 commented:
Ok I think I've now addressed all the comments hopefully, although there were a lot of formatting changes so hopefully didn't introduce too many new problems..

sibirrer commented on lenstronomy/Workflow/psf_fitting.py:
```diff
@@ -38,15 +39,38 @@ class to find subsequently a better psf
+        calculate the completeness numerically when rotational symmetry is imposed. This is the simplest 'mask' which throws away
+        anywhere the rotations are not fully complete ->e.g. in the corners. This ONLY accounts for information loss in
+        corners, not due e.g. to losses at the edges of the images.
+        :param kwargs_psf: all the psf_kwargs, really only the kernel size is needed here
```
add space between description and the input definitions.

test ...

:param ....

sibirrer commented on lenstronomy/Workflow/psf_fitting.py:
```diff
@@ -79,8 +103,17 @@ def update_iterative(self, kwargs_psf, kwargs_params, num_iter=10, keep_psf_erro
+
         for i in range(num_iter):
-            kwargs_psf_new, logL_after, error_map = self.update_psf(kwargs_psf_new, kwargs_params, **kwargs_psf_update)
+            kwargs_psf_new, logL_after, error_map = self.update_psf(kwargs_psf_new, kwargs_params, corner_mask,**kwargs_psf_update)
```
have a space between 'corner_mask, **kwargs_psf_update'

sibirrer commented on test/test_Workflow/test_psf_fitting.py:
```diff
@@ -113,6 +113,24 @@ def test_update_psf(self):
         diff_new = np.sum((kernel_new - kernel_true) ** 2)
         assert diff_old > diff_new
 
```
add a separate test function for calc_cornermask() definition

sibirrer commented on test/test_Workflow/test_psf_fitting.py:
```diff
@@ -161,7 +179,6 @@ def test_mask_point_source(self):
         mask_point_source = self.psf_fitting.mask_point_source(ra_image, dec_image, x_grid, y_grid, radius, i=0)
         assert mask_point_source[10, 10] == 1
 
-
```
add line. Two lines separation outside classes

sibirrer commented on test/test_Workflow/test_psf_fitting.py:
```diff
@@ -299,6 +316,5 @@ def test_update_iterative(self):
         assert diff_old > diff_new
         assert diff_new < 0.01
 
-
```
add line. Two lines separation outside classes

sibirrer reviewed (changes requested):
Looks good. I have one larger request to add a test function for the calc_cornermask definition function and some minor pedantic things.
Thank you very much!

amn3142 commented:
Ok hopefully this time I caught everything sorry I missed those formatting errors that you had already pointed out :(

sibirrer reviewed (approved):
This looks great. Thank you very much!
